File: week1/credit.c
#include <cs50.h>
#include <stdio.h>
int main(void)
{
    long n;
    long productDigitSum = 0;
    long productSum = 0;
    long sumOfNotMultiplier = 0;
    long check;
    bool creditCardIsValid = false;

    while (!creditCardIsValid) {
        do {
            n = get_long("please enter your credit card number: \n");
        }
        while (n < 1);
        long save_n = n;
        long save_n_2 = n;
        long save_n_3 = n;
        long save_n_4 = n;
        long save_n_5 = n;
        n = n / 10;
        do {
            productSum = productSum + (n % 10) * 2;
            if (((n % 10) * 2) < 10) {
                productDigitSum = productDigitSum + (n % 10) * 2;
            }
            else {
                productDigitSum = productDigitSum + (((n % 10) * 2) % 10) + ((((n % 10) * 2) / 10) % 10);
            }
            n = n / 100;
        }
        while (n > 0 || (n / 10) > 10);

        do {
            sumOfNotMultiplier = sumOfNotMultiplier + (save_n % 10);
            save_n = save_n / 100;
        }
        while(save_n > 0 || (save_n / 100) >= 10);

        check = sumOfNotMultiplier + productDigitSum;
        if (check % 10 == 0) {
            int i = 0;
            while (save_n_2 > 0) {
                save_n_2 = save_n_2 / 10;
                i++;
            }
                if (i == 15) {
                    while (save_n_3 > 100) {
                        save_n_3 = save_n_3 / 10;
                    }
                    if (save_n_3 == 34 || save_n_3 == 37) {
                        printf("AMEX\n");
                        creditCardIsValid = true;
                    }
                    else if ((save_n_3 / 10) == 4) {
                        printf("VISA\n");
                        creditCardIsValid = true;
                    }
                    else {
                        printf("Unknown credit id, ");
                    }
                }
                else if (i == 16) {
                    while (save_n_4 > 100) {
                        save_n_4 = save_n_4 / 10;
                    }
                    if (save_n_4 == 51 || save_n_4 == 52 || save_n_4 == 53 || save_n_4 == 54 || save_n_4 == 55) {
                        printf("MASTER\n");
                        creditCardIsValid = true;
                    }
                    else if ((save_n_4 / 10) == 4) {
                        printf("VISA\n");
                        creditCardIsValid = true;
                    }
                    else {
                        printf("Unknown credit id, ");
                    }
                }
                else if (i == 13 || i == 14) {
                    while (save_n_5 > 100) {
                        save_n_5 = save_n_5 / 10;
                    }
                    if ((save_n_5 / 10) == 4) {
                        printf("VISA\n");
                        creditCardIsValid = true;
                    }
                    else {
                        printf("Unknown credit id, ");
                    }
                }
                else {
                    printf("Unknown credit id, ");
                }
            }
        else {
            printf("INVALID\n");
        }
    }
}

File: week6/sentimental-credit/credit.py
import re

def main():
    credit_card_is_valid = False
    while not credit_card_is_valid:
        individual_digit_of_product_sum = 0
        sum_of_not_multiplier = 0
        credit_card_number = int(input("please enter your credit card number:  "))
        credit_card_number_copy = credit_card_number
        credit_card_number_copy = int(credit_card_number_copy / 10)
    
        while True:
            if ((credit_card_number_copy  % 10) * 2) < 10 :
                individual_digit_of_product_sum = individual_digit_of_product_sum + (credit_card_number_copy % 10) * 2
            else: 
                individual_digit_of_product_sum = individual_digit_of_product_sum + (((credit_card_number_copy % 10) * 2) % 10) + int((((credit_card_number_copy % 10) * 2) / 10)) % 10

            if not credit_card_number_copy >= 0 or not (int(credit_card_number_copy) / 10) >= 10:
                break

            credit_card_number_copy = int(credit_card_number_copy / 100)

        credit_card_number_copy = credit_card_number

        while True:
            sum_of_not_multiplier = sum_of_not_multiplier + (credit_card_number_copy % 10)
            credit_card_number_copy = int(credit_card_number_copy / 100)

            if not (credit_card_number_copy > 0 or (int(credit_card_number_copy / 100)) > 10):
                break

        check = sum_of_not_multiplier + individual_digit_of_product_sum
        check_is_valid = (check % 10) == 0
        credit_card_number = str(credit_card_number)

        if check_is_valid:
            if (len(credit_card_number) == 15):
                is_amex = re.findall("^34", credit_card_number) or re.findall("^37", credit_card_number)
                if is_amex:
                    print("AMEX")
                    credit_card_is_valid = True
                else:
                    print("Unknown credit id, ")

            elif (len(credit_card_number) == 16):
                is_master = re.findall("^51", credit_card_number) or re.findall("^52", credit_card_number) or re.findall("^53", credit_card_number) or re.findall("^54", credit_card_number) or re.findall("^55", credit_card_number)
                is_visa = re.findall("^4", credit_card_number)
                if is_master:
                    print("MASTER")
                    credit_card_is_valid = True
                elif is_visa:
                    print("VISA")
                    credit_card_is_valid = True
                else:
                    print("Unknown credit id, ")

            elif (len(credit_card_number) == 13 or len(credit_card_number) == 14):
                is_visa = re.findall("^4", credit_card_number)
                if is_visa:
                    print("VISA")
                    credit_card_is_valid = True
                else:
                    print("Unknown credit id, ")
                    
        else:
            print("INVALID")

main()
